### PR TITLE
feat: add new binding type attribute to BPMN model API

### LINKED ISSUE #19461: [`deployment` version binding] Add new `bindingType` attribute to BPMN Model API
### Description
Add a new (optional) `bindingType` attribute to the `ZeebeCalledElement`, `ZeebeCalledDecision`, and `ZeebeFormDefinition` extension elements in the BPMN Model API.

### Acceptance Criteria
- [ ] A new attribute `bindingType` is added to `ZeebeCalledElement`, `ZeebeCalledDecision`, and `ZeebeFormDefinition`.
  - [ ] The attribute is optional.
  - [ ] Allowed values are `deployment` and `latest`.
  - [ ] `latest` is used as the default if the attribute is not set.
- [ ] The attribute can be set via the respective builder classes of the parent elements:
  - [ ] `AbstractBusinessRuleTaskBuilder`
  - [ ] `AbstractCallActivityBuilder`
  - [ ] `AbstractStartEventBuilder`
  - [ ] `AbstractUserTaskBuilder`

### PR DESCRIPTION
**Pull request by jfriedenstab**
## Description
Adds support for a new `bindingType` attribute (that can be `latest` or `deployment`) to the `ZeebeCalledElement`, `ZeebeCalledDecision`, and `ZeebeFormDefinition` extension elements in the BPMN Model API.
This should allow Zeebe to determine the binding type for such elements when a deployed BPMN diagram is processed.

## Related issues

closes #19461

### CODE REVIEW

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java
Comment (jfriedenstab) [inline zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java:180]: FYI: Given that you can set a couple of other attributes that are not related to the form, I named this method `zeebeFormBindingType` instead of just `zeebeBindingType`.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java
Comment (jfriedenstab) [inline zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java:109]: I decided not to set a default value here, because the binding type is not applicable when using `formKey` or `externalReference`.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/ZeebeFormDefinitionValidator.java
Comment (jfriedenstab) [inline zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/ZeebeFormDefinitionValidator.java:70]: I'm not sure whether such a validation is actually needed (the `bindingType` could also just be ignored if it was used with a `formKey` or `externalReference`). But I think it doesn't hurt to be strict here and return an error.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java
Comment (korthout) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java:180]: 💭 It would be good to [document such intentions in the commit message](https://github.com/camunda/camunda/wiki/Pull-Requests-and-Code-Reviews#ive-written-a-clear-changelist-description) as well. This makes it easier to re-discover such decisions later, as the git commit message are available locally close to the code in most editors compared to GitHub PR comments which exist on the web. [Splitting up your commit into multiple smaller ones](https://github.com/camunda/camunda/wiki/Pull-Requests-and-Code-Reviews#ive-narrowly-scoped-my-changes) can help highlight changes for which a reason should be documented.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeBindingType.java
Comment (korthout) [inline zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeBindingType.java:20]: 💭 I was surprised to see non-capitalized enum values, but they appear to be needed:
https://github.com/camunda/camunda/pull/16519#discussion_r1508047718 TIL

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java
Comment (korthout) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java:109]: Not setting a default means that existing models with a FormDefinition using `formId` (those without `bindingType`) are implicitly using `latest` even though there is no `bindingType` specified and no default either. I'm not sure whether it's a big problem, but perhaps we see a better solution. 

❓ Do you see any way to overcome this?

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/ZeebeFormDefinitionValidator.java
Comment (korthout) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/ZeebeFormDefinitionValidator.java:70]: 🔧 With my previous comment in mind, it may be better to allow `bindingType` and `formKey` at the same time, to allow a default `bindingType` value.

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/builder/CallActivityBuilderTest.java
Comment (korthout) [inline zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/builder/CallActivityBuilderTest.java:53]: 👍 Thanks for adding additional coverage 🙇

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ZeebeTaskValidatorFormTest.java
Comment (korthout) [inline zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ZeebeTaskValidatorFormTest.java:364]: 👍 I do like that we're helping users by clear rejection messages

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ZeebeTaskValidatorFormTest.java
Comment (korthout) [inline zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ZeebeTaskValidatorFormTest.java:655]: 🔧 We could consider adding a test that shows what happens when a string is used that is not an enum value of `ZeebeBindingType` as `bindingType`

**Review by korthout — COMMENTED** [review]
Nice work @jfriedenstab 👏 

👍 I like the changes and choices you've made

❌ There's one open question that we must discuss before I can approve the changes. Please have a look.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java:180]: Thanks, good points! 👍🏻 We're not really using the commit message body in my team, so I still need to get used to providing more information there ;) I've also realized that I could have split up my commit into multiple smaller ones. From this repo's history I see that the PRs often contain relatively small individual commits.

I will keep the points in mind for the next PR. Or was your comment meant as a suggestion to rewrite the commit history?

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeBindingType.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeBindingType.java:20]: I actually took `ZeebeExecutionListenerEventType` as an inspiration 🙂. But, yes, it looks a bit unusual indeed.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java:109]: Thanks for raising this 👍🏻.

> Not setting a default means that existing models with a FormDefinition using `formId` (those without `bindingType`) are implicitly using `latest` even though there is no `bindingType` specified and no default either.

Hm, I think the problem is rather that for these models the `bindingType` will be `null` and not `latest`, right? However, `latest` would make sense as this is already the current behavior in Zeebe if a `formId` is used.

I would basically see three options to deal with this:
1. Always use `latest` as the default if the `bindingType` is not explicitly set in the model, even if `formKey` or `externalReference` is used. This would be the simplest solution (and also resolve your [other comment](https://github.com/camunda/camunda/pull/19766#discussion_r1658848845)).

2. Use `latest` as the default if the `bindingType` is not explicitly set in the model, but _only_ if the `formId` is set (probably by enhancing the `getBindingType` method). This would add a bit of extra complexity, but reflect my [original intention](https://github.com/camunda/camunda/pull/19766#discussion_r1655076741).

3. Keep the current implementation. This would require an extra `null` check in Zeebe later when the binding type is actually used to resolve the form by `formId` (`null` would be equivalent to `latest` then).

I would prefer 1 or 2, also to be in line with `ZeebeCalledElementImpl` and `ZeebeCalledDecisionImpl` which actually define a default value.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/ZeebeFormDefinitionValidator.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/ZeebeFormDefinitionValidator.java:70]: Yes, if we decide to always set a default value for `bindingType`, let's just remove this extra validation.

One use case I had in mind when adding it was that users might have specified the wrong attribute by mistake (e.g. they actually wanted to use `formId` with `bindingType=deployment`, but used `formKey` instead). But that would probably be quite an edge case.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java
Comment (korthout) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java:180]: @jfriedenstab My intention was simply to point it out as something for future changes. However, if you think it's easy to change then please take it along.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java
Comment (korthout) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java:109]: It seems to me that semantically it's not incorrect to use `latest` with `formKey`. There simply is only one version of it. 🤷 

I'm not a fan of unnecessary complexity. It doesn't appear to me as something that significantly helps our users. If it does, we could always add such rules later as improvements.

Given that option 1 is the simplest, it has my preference until we see a user need.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java:180]: It was relatively easy, so I split the commit into two.

File: zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/impl/instance/zeebe/ZeebeFormDefinitionImpl.java:109]: Let's go with option 1 then, thanks! :)

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ZeebeTaskValidatorFormTest.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ZeebeTaskValidatorFormTest.java]: Good suggestion 👍🏻. I added such tests with https://github.com/camunda/camunda/pull/19766/commits/9efc3ee74a44974b1ce19e615a2dc08b881c087c.

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java
Comment (jfriedenstab) [inline zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java:84]: 💭 Maybe it would make sense to add a validation rule to return a more user-friendly error if an invalid binding type was used (something like `Attribute 'bindingType' must be one of: deployment, latest`). But I would add this in a follow-up PR.

**Comment by jfriedenstab** [comment]
@korthout I addressed all the comments. Could you have another look again?

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java
Comment (korthout) [inline zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java:56]: 🔧 From the name of the test method and from the name of the bpmn resource it's unclear that this verifies that the binding type of the called decision is `deployment`. My suggestion is to adjust the bpmn filename.

🔧 I'm not a big fan of bpmn files in tests, as they make the tests harder to understand (the information is split over multiple places), like above. We could use the bpmn-model api to create the bpmn model, produce the xml, and then read it again all in the same test case to avoid this.

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java
Comment (korthout) [reply zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java:84]: 👍 I like this idea, and would suggest we take this along while this is still fresh. Fine to do so in a separate PR though.

**Review by korthout — APPROVED** [review]
Looking great @jfriedenstab 🚀 

🔧 I have a few last suggestions, but nothing blocking this from being merged when you see fit. Feel free to apply the suggestions or just merge.

👍 LGTM

💭 Sorry for the delayed review

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java]: Thanks, good points! I replaced the BPMN files with models created by the API for both cases (valid and invalid binding type).

File: zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java
Comment (jfriedenstab) [reply zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/instance/zeebe/ZeebeCalledDecisionTest.java:100]: I created a [follow-up issue](https://github.com/camunda/camunda/issues/20054) for this and will tackle it soon.

**Pull request merged**
